### PR TITLE
Async

### PR DESCRIPTION
**Pull request by stelcheck**
- Makefile bugfix
- Async calls (either deferred or truly async, configurable)

**Todo**
- [ ] Test on Android and iOS
- [ ] Test on cocos2dx

@tokutake @MiLk @aogilvie would love if one of you (or all!) would be able to take some time and test this on the environments you care about. I will also be doing some tests on my side. Let me know if you find anything broken in the meantime

### CODE REVIEW

**Comment by stelcheck** [comment]
For Android, I am getting this issue https://github.com/mage/mage-sdk-cpp/issues/17, but also:

```
/Users/mt/android-ndk/sources/cxx-stl/gnu-libstdc++/4.6/include/future:111:11: error: declaration of 'struct std::future<Json::Value>'
/Users/mt/Sources/mage-sdk-cpp/platforms/android/jni/../../../src/mage.cpp:49:22: error: 'deferred' is not a member of 'std::launch'
/Users/mt/Sources/mage-sdk-cpp/platforms/android/jni/../../../src/mage.cpp: In member function 'void mage::RPC::SetDomain(std::string)':
/Users/mt/Sources/mage-sdk-cpp/platforms/android/jni/../../../src/mage.cpp:61:3: error: 'httpClient' was not declared in this scope
/Users/mt/Sources/mage-sdk-cpp/platforms/android/jni/../../../src/mage.cpp: In member function 'void mage::RPC::SetApplication(std::string)':
/Users/mt/Sources/mage-sdk-cpp/platforms/android/jni/../../../src/mage.cpp:66:3: error: 'httpClient' was not declared in this scope
/Users/mt/Sources/mage-sdk-cpp/platforms/android/jni/../../../src/mage.cpp: In member function 'void mage::RPC::SetProtocol(std::string)':
/Users/mt/Sources/mage-sdk-cpp/platforms/android/jni/../../../src/mage.cpp:71:3: error: 'httpClient' was not declared in this scope
/Users/mt/Sources/mage-sdk-cpp/platforms/android/jni/../../../src/mage.cpp: In member function 'void mage::RPC::SetSession(std::string)':
/Users/mt/Sources/mage-sdk-cpp/platforms/android/jni/../../../src/mage.cpp:75:3: error: 'httpClient' was not declared in this scope
/Users/mt/Sources/mage-sdk-cpp/platforms/android/jni/../../../src/mage.cpp: In member function 'void mage::RPC::ClearSession()':
/Users/mt/Sources/mage-sdk-cpp/platforms/android/jni/../../../src/mage.cpp:79:3: error: 'httpClient' was not declared in this scope
/Users/mt/android-ndk/sources/cxx-stl/gnu-libstdc++/4.6/include/future: At global scope:
/Users/mt/android-ndk/sources/cxx-stl/gnu-libstdc++/4.6/include/future:143:5: error: 'std::future<typename std::result_of<_Functor(_ArgTypes ...)>::type> std::async(std::launch, _Fn&&, _Args&& ...) [with _Fn = mage::RPC::Call(const string&, const Json::Value&, bool)::<lambda()>, _Args = {}, typename std::result_of<_Functor(_ArgTypes ...)>::type = Json::Value]', declared using local type 'mage::RPC::Call(const string&, const Json::Value&, bool)::<lambda()>', is used but never defined [-fpermissive]
```

**Comment by stelcheck** [comment]
iOS suffers also from #17 but else compiles successfully.

**Comment by stelcheck** [comment]
The last commit should fix it for Android. Once #17 is fixed I expect things will work properly, but will still need to do more tests.

**Comment by stelcheck** [comment]
Now also fixes #17 - thanks to @tokutake for the help.

The only things left to test should be integration and the async behaviour. @tokutake would this PR solve your issues when it comes to async calls?

**Comment by stelcheck** [comment]
I just wrote some small test to verify the two async behaviours we might want to use, and both do work properly. So I expect there should be no problem under other platforms once we get to confirm that this PR does work for them.

File: src/mage.cpp [hunk 1]
```diff
@@ -40,6 +40,18 @@ namespace mage
 		return res;
 	}
 
+	std::future<Json::Value> RPC::Call(const std::string &name, const Json::Value &params, bool doAsync) {
+		if (doAsync) {
```
Comment (MiLk) [inline src/mage.cpp]: What about the following?

``` c++
        std::launch policy = doAsync ? std::launch::async : std::launch::deferred;
        return std::async(policy, [this, name, params]{
            return Call(name, params);
        });
```

Comment (stelcheck) [inline src/mage.cpp]: Yup, that is better, ill do that.

**Comment by MiLk** [comment]
What problem did you try to solve with the async behaviour?

[std::future::get](http://en.cppreference.com/w/cpp/thread/future/get) is blocking, I don't really see a change between `examples/simple` and `examples/async` (with `doAsync` set to true).
The call will be run in the separated thread, but the main thread will wait until the new thread ends.
The only case, where I think it's useful, is to run multiple call in parallel and use [std::future::wait](http://en.cppreference.com/w/cpp/thread/future/wait).

If you need something really async, you'll probably have to use `std::async` directly in your application.

**Comment by stelcheck** [comment]
get is blocking, but the computation of the call is not. This means that you will, if setting doAsync to true, make the call to MAGE without blocking your main thread.

I am assuming this is all that would be needed for the purpose of making stuff such as loading bars, etc, but arguably since calls happen in parallel it would be nice to get notified when they are done instead of not knowing the state of a call. If you have any suggestions, please let me know.

**Pull request merged**